### PR TITLE
Addition of Item Actions and Update od UplaodHelper

### PR DESCRIPTION
**Pull request by dvnstrcklnd**

### CODE REVIEW

File: standard_libs/libraries/itemactions/source.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+module ItemActions
+  # Store all items used in input operations
+  # Assumes all inputs are non nil
+  #
+  # @param operations [OperationList] the list of operations
+  # @param location [String] the storage location
+  # @param type [String] the type of items to be stored ('item', 'collection', nil)
+  def store_inputs(operations, location: nil, type: nil)
+    store_io(operations, role: 'inputs', location: location, type: type)
```
Comment (bjkeller) [inline standard_libs/libraries/itemactions/source.rb]: I think `role` should match field value role, which is singular

File: standard_libs/libraries/itemactions/source.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        raise 'Invalid role'
+      elsif role.downcase == 'inputs'
+        field_values = op.inputs
+      elsif role.downcase == 'outputs'
+        field_values = op.outputs
+      else
+        field_values = op.outputs + op.inputs
+      end
+
+      unless type.nil?
```
Comment (bjkeller) [inline standard_libs/libraries/itemactions/source.rb:47]: why would you only want to store some kinds of I/O?

BTW the canonical way to check if a field value is a collection is to check the object type handler for the item.
The code written here seems potentially flakey.

If this code is necessary, I would add a helper to this library 

```ruby
def collection?(item)
  item.object_type.handler == 'collection'
end
```

and then call it instead.

Newer versions of Aquarium also have a predicate `fv.collection?`, but you can't yet count on systems being up to date.

File: standard_libs/libraries/itemactions/source.rb [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    tab = create_location_table(items)
+    show do
+      title 'Properly Dispose of the following items:'
+      table tab
+    end
+    items.each {|item| item.mark_as_deleted}
+  end
+    
+    
+end
```
Comment (bjkeller) [inline standard_libs/libraries/itemactions/source.rb]: add a newline

has this code been passed through rubocop?

File: standard_libs/libraries/itemactions/definition.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+  "id": 19,
```
Comment (bjkeller) [inline standard_libs/libraries/itemactions/definition.json:2]: I don't think pfish is using the ID anymore.  I would doublecheck and the remove it.

File: standard_libs/libraries/uploadhelper/definition.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
-  "id": 19,
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/definition.json:2]: lose the ID

File: standard_libs/libraries/itemactions/source.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+  # Stores all items of a certain role in the operations list
+  # Creates instructions to store items as well
+  #
+  # @param operations [OperationList] list of Operations
+  # @param role [String] whether material to be stored is an input or an output
+  # @param location [String] the location to store the material
+  # @param all_items [Boolean] an option to store all items not just collections
+  # @param type [String] the type of items to be stored ('item', 'collection', nil)
+  def store_io(operations, role: 'all', location: nil, type: nil)
```
Comment (bjkeller) [inline standard_libs/libraries/itemactions/source.rb:36]: what is the behavior when `location` is `nil`?

File: standard_libs/libraries/itemactions/source.rb [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        if type.downcase == 'collection'
+          field_values.reject { |fv| fv.collection.nil? }
+        elsif type.downcase == 'item'
+          field_values.reject { |fv| !fv.collection.nil? }
+        end
+      end
+
+      items.concat(field_values.map(&:item))
+    end
+    items.uniq!
```
Comment (bjkeller) [inline standard_libs/libraries/itemactions/source.rb]: I haven't looked, but I think if you can do `items.uniq!`, then `items` could be a `Set` and you get the uniqueness for "free".

File: standard_libs/libraries/uploadhelper/source.rb [hunk 8]
```diff
@@ -1,87 +1,161 @@
...
+  # @param dirname - directory where files are located, or full path including filename
+  # @param exp_upload_num - expected number of files to upload
+  # @param tries - max. number of attempts to upload expectedNum files
+  #
+  # @returns array of Upload items
+  #
+  # EXAMPLES of how to associate correctly:
+  # data associations - 1st gel image
+  # up_bef=ups[0]
+  # op.plan.associate 'gel_image_bef', 'combined gel fragment', up_bef  # upload association, link
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: use parentheses in example code

File: standard_libs/libraries/uploadhelper/source.rb [hunk 9]
```diff
@@ -1,87 +1,161 @@
...
+  # op.input(INPUT).item.associate 'gel_image_bef', up_bef              # regular association
+  # op.output(OUTPUT).item.associate 'gel_image_bef', up_bef            # regular association
+  #------------------------------------------
+  def upload_data(dirname, exp_upload_num, tries)
+    uploads = {} # result of upload block
+    number_of_uploads = 0 # number of uploads in current attempt
+    attempt = 0 # number of upload attempts
+
+    loop do
+      # if(number_of_uploads==exp_upload_num)
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: remove commented out code

File: standard_libs/libraries/uploadhelper/source.rb [hunk 10]
```diff
@@ -1,87 +1,161 @@
...
+      # show {note 'Upload complete.'}
+      # end
+      break if (attempt >= tries) || (number_of_uploads == exp_upload_num)
+
+      attempt += 1
+
+      uploads = show do
+        title "Select <b>#{exp_upload_num}</b> file(s)"
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: these instructions seem very nonspecific: "select 5 files"

File: standard_libs/libraries/uploadhelper/source.rb [hunk 11]
```diff
@@ -1,87 +1,161 @@
...
+  def upload_data(dirname, exp_upload_num, tries)
+    uploads = {} # result of upload block
+    number_of_uploads = 0 # number of uploads in current attempt
+    attempt = 0 # number of upload attempts
+
+    loop do
+      # if(number_of_uploads==exp_upload_num)
+      # show {note 'Upload complete.'}
+      # end
+      break if (attempt >= tries) || (number_of_uploads == exp_upload_num)
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: is there a reason this shouldn't be a while-loop?

File: standard_libs/libraries/uploadhelper/source.rb [hunk 12]
```diff
@@ -1,87 +1,161 @@
...
         end
-        
-        # return
-        ups
-        
-    end # def
+        upload var: 'files'
+      end
+      # number of uploads
+      if !uploads[:files].nil?
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: `number_of_uploads = uploads[:files].length unless uploads[:files].nil?`

File: standard_libs/libraries/uploadhelper/source.rb [hunk 13]
```diff
@@ -1,87 +1,161 @@
...
+    end
+
+    if number_of_uploads != exp_upload_num
+      show {note "Final number of uploads (#{number_of_uploads}) not equal to expected number #{exp_upload_num}! Please check."}
+      return nil
+    end
+
+    # format uploads before returning
+    ups = [] # upload hashes
+    if !uploads[:files].nil?
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: eliminate the nested conditional:

`return nil if uploads[:files].nil?`

File: standard_libs/libraries/uploadhelper/source.rb [hunk 14]
```diff
@@ -1,87 +1,161 @@
...
+  def read_url(upload)
+    url = upload.url
+    matrix = []
+    CSV.new(open(url)).each { |line| matrix.push(line) }
+    # open(url).each { |line| matrix.push(line.split(',') }
+  end
+
+  # Validates upload and ensures that it is correct
+  #
+  # @param uplaod_array Array array of  uploads
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: "uplaod"

File: standard_libs/libraries/uploadhelper/source.rb [hunk 15]
```diff
@@ -1,87 +1,161 @@
...
+  # @param upload [upload_obj] the file that you wish to read from
+  # @return matrix [2D-Array] is the array of arrays of the rows read from file, if csv
+  def read_url(upload)
+    url = upload.url
+    matrix = []
+    CSV.new(open(url)).each { |line| matrix.push(line) }
+    # open(url).each { |line| matrix.push(line.split(',') }
+  end
+
+  # Validates upload and ensures that it is correct
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: what does this mean?

File: standard_libs/libraries/uploadhelper/source.rb [hunk 16]
```diff
@@ -1,87 +1,161 @@
...
+        note "Please hit <b>Okay</b> to try again"
+      end
+      return false
+    end
+
+    fail_message = ''
+
+
+    # TODO this makes no sense....   Deff should expect more than one file sometimes...
+    if multiple_files == false
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: `if !multiple_files`

File: standard_libs/libraries/uploadhelper/source.rb [hunk 17]
```diff
@@ -1,87 +1,161 @@
...
+      fail_message += 'More than one file was uploaded, ' if upload_array.length > 1
+      upload = upload_array.first
+    end
+
+    csv = CSV.read(open(upload.url))
     
+    fail_message += 'CSV length is shorter
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: what is the length of a csv?  this is number of rows, right?

File: standard_libs/libraries/uploadhelper/source.rb [hunk 18]
```diff
@@ -1,87 +1,161 @@
...
-    # @param upload [upload_obj] the file that you wish to read from
-    # @return matrix [2D-Array] is the array of arrays of the rows read from file, if csv
-    def read_url(upload)
-        url = upload.url
-        matrix = []
-        CSV.new(open(url)).each {|line| matrix.push(line)}
-        # open(url).each {|line| matrix.push(line.split(',')}
-        return matrix
+    
+    # TODO Make this more robust.   Sometimes CSV files contain unprintable characters
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: is this an encoding issue, or a character issue?

File: standard_libs/libraries/uploadhelper/source.rb [hunk 19]
```diff
@@ -1,87 +1,161 @@
...
 
-   
\ No newline at end of file
+  # Needs documentation
+  def get_validated_uploads(expected_data_points, headers, multi_files, file_location: 'Unknown Location',
+          detailed_instructions: nil)
+    tries = 1
+    max_tries = 10
+    pass = false
+    until pass == true
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: `until pass`

File: standard_libs/libraries/uploadhelper/source.rb [hunk 20]
```diff
@@ -1,87 +1,161 @@
...
+                detailed_instructions: detailed_instructions)
+      pass = validate_upload(csv_uploads, expected_data_points, headers, multiple_files: multi_files)
+      tries += 1
+      raise 'Too many failed upload attempts' if tries == max_tries && !pass
+    end
+    csv_uploads
+  end
+
+  # Instructions to upload CSV files of concentrations
+  def upload_csv(tries = nil, max_tries = 'NA', file_location: 'Unknown Location',
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: other than being one file, why is this different than `upload_data`?

File: standard_libs/libraries/uploadhelper/source.rb [hunk 21]
```diff
@@ -1,87 +1,161 @@
...
+    # open(url).each { |line| matrix.push(line.split(',') }
+  end
+
+  # Validates upload and ensures that it is correct
+  #
+  # @param uplaod_array Array array of  uploads
+  # @param expected_num_inputs int the expected number of inputs
+  # @param csv_headers array array of expected headers
+  # @returns pass Boolean pass or fail (true is pass)
+  def validate_upload(upload_array, expected_num_inputs, csv_headers, multiple_files: true)
```
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: make a method for each check

File: standard_libs/libraries/uploadhelper/source.rb (hunk 20 above)
Comment (bjkeller) [inline standard_libs/libraries/uploadhelper/source.rb]: `tries` and `max_tries` defaults should be numbers

**Review by bjkeller — COMMENTED** [review]
- use more guards to reduce nested conditionals
- some of the parameter lists seem extra long
- some of the defaults for parameters seem incorrect (wrong value or wrong type)
- upload methods seem to be redundant and validation message code is awkward

**Comment by dvnstrcklnd** [comment]
@malloc3 can you address these issues?

File: standard_libs/libraries/itemactions/source.rb (hunk 6 above)
Comment (malloc3) [reply standard_libs/libraries/itemactions/source.rb:36]: It will not set the location to anything.   The old location shall be retained

**Comment by malloc3** [comment]
Issues with ItemActions have been addressed.   UploadHelper has been removed from this pull request and will be addressed at a later time. @bjkeller @dvnstrcklnd

**Comment by dvnstrcklnd** [comment]
@malloc3 @bjkeller this one slipped off my radar. Do we still need to merge the changes to ItemActions? If so, this commit seems to also have too many files being changed.

**Comment by dvnstrcklnd** [comment]
@bjkeller have your concerns been resolved? If we merge this, will it delete a bunch of files? I am not sure how that works out.

**Comment by malloc3** [comment]
I am closing this pull request.   Its old and not sure it is correct anymore.  I will create a new branch for only ItemActions that should be cleaner and will avoid concerns of files getting deleted

**Comment by malloc3** [comment]
Correction Item actions was already pushed and merged to master.  THis can be ignored completely.

**Pull request closed without merging**